docs/devel/api-conventions.md

```diff
@@ -370,6 +371,25 @@ All compatible Kubernetes APIs MUST support "name idempotency" and respond with
 
 Names generated by the system may be requested using `metadata.generateName`. GenerateName indicates that the name should be made unique by the server prior to persisting it. A non-empty value for the field indicates the name will be made unique (and the name returned to the client will be different than the name passed). The value of this field will be combined with a unique suffix on the server if the Name field has not been provided. The provided value must be valid within the rules for Name, and may be truncated by the length of the suffix required to make the value unique on the server. If this field is specified, and Name is not present, the server will NOT return a 409 if the generated name exists - instead, it will either return 201 Created or 504 with Reason `ServerTimeout` indicating a unique name could not be found in the time allotted, and the client should retry (optionally after the time indicated in the Retry-After header).
 
+## Optional vs Required
+
+Fields must be either optional or required.
+
+Optional fields have the following properties:
+
+- They have `omitempty` struct tag in Go.  This causes swagger documentation to reflect that the field is empty.
+- They are a pointer type in the Go definition (e.g. `bool *awesomeFlag`).  Using a pointer allows distinguishing unset from the default value.
```

Comment:
Do we really want to make all optional fields as pointers?
I thought we make an optional field pointer, only when that type's default value is a valid user-specified value and that is not the default value we want. 
For ex: ReplicationController.Replicas is a pointer since we want to know if the user specified it as 0 or its value is 0 because default value of int is 0. If its not specified by user, then we want to change that to 1.
ReplicationController.RestartPolicy is not a pointer since we know that "" is not a valid user specified value for that field. We can set it to appropriate default, when it is "".
ReplicationController.Volumes is not a pointer since the default value (an empty array) is sufficient.